### ISSUE TITLE
Reparametrization in VGAE model can cause NAN's

### ISSUE THREAD

**Issue by joeybose** [rusty1s/pytorch_geometric#500]
## ❓ Questions & Help

While, the implementation of the reparametrization is clean and conventional it can potentially lead to `NaN's` as the `self.__logvar__` term is not controlled. This behavior is dependent on the use case, and for the most part will not cause problems for many applications but I believe a better fix could be something like to add another line after : https://github.com/rusty1s/pytorch_geometric/blob/master/torch_geometric/nn/models/autoencoder.py#L231 
```
 self.__logvar__ = torch.clamp(self.__logvar__,min=LOG_VAR_MIN,max=LOG_VAR_MAX)
```
Where, `LOG_VAR_MIN` and `LOG_VAR_MAX` are suitable constants. This prevents numerical issues (underflow/overflow) that can happen during optimization.

**Comment by rusty1s** [comment]
Thank you. I didn‘t know this is common in practice. So i guess it is not the encoder producing NaNs but the `exp` calls afterwards. Wouldn‘t it be better to just filter out the NaNs after those calls?

**Comment by joeybose** [comment]
One can filter out NaNs after the fact as well, I've been using the above trick and it seems to work but maybe its more elegant to filter out NaN's? I think both methods should work.

**Comment by rusty1s** [comment]
So what do you use for `LOG_VAR_MIN` and `LOG_VAR_MAX`?

**Comment by joeybose** [comment]
I've been using `LOG_VAR_MIN=2` and `LOG_VAR_MAX=20` and it was basically taken from another repo and it seems to work. There may be better constants, but for my use case, it solved the issue.

**Comment by rusty1s** [comment]
But this would result in zero gradients if the encoder does not produce `logvar` in this interval. I can see the reasoning behind clamping high values, but why should we clamp values smaller than 2? There shouldn't be any numerical issues if I am not mistaken.

**Comment by joeybose** [comment]
Well, most of the numerical instability comes from values being too close to 0, so there should be a clamp from below.  I agree however that 2 might be too high and can be lower.

**Comment by rusty1s** [comment]
But why? There are no instabilities when taking `exp(0)`?

**Comment by joeybose** [comment]
Well, the output from a linear layer is unbounded so if you have a large negative number than `exp()` of that could be numerically unstable.

**Comment by rusty1s** [comment]
Yes, it may underflow and result in zeros, but certainly not in NaNs, or what am I missing? Can you do me a favor and test if your model still works if you only clamp high values.

**Comment by joeybose** [comment]
Sure I'll update this tonight with the result.

**Comment by joeybose** [comment]
Seems like your right, and you don't need `LOG_VAR_MIN`. Only having `LOG_VAR_MAX` solves the issue (atleast in my use case)!

**Comment by rusty1s** [comment]
Perfect, will add it :)

**Comment by joeybose** [comment]
Small correction `LOG_VAR_MAX=20` blows up still so a lower value like `10` is more stable. But other than that everything is as you predicted.

**Comment by rusty1s** [comment]
Added to master. Thank you!
